### PR TITLE
Add Google Sheets Output

### PR DESCRIPTION
**Pull request by danieldortega77**
The module now can:

- Read from all cells in a Google Sheet, including all subsheets within
- Creates a new spreadsheet named by the user's input
- Appends Perspective analysis to the new sheet
- Color codes the data in the new sheet

Screenshot 2020-08-03 at 10 07 42 AM - Display 1

### CODE REVIEW

File: capstone/src/main/webapp/sheetsScript.js [hunk 1]
```diff
@@ -62,29 +64,272 @@ async function gatherSheetsInput() {
...
+  let totalText = '';
+  let range = '!A1:YY';
+  for (const name of sheetNames) {
+    const sheet = await getSpreadsheet(id, name, range);
+    const text = await getTextFromSheet(sheet);
+    totalText += text + '\n';
+  }
+  handleInput(totalText, langElement.value, requestedAttributes, tokenizer);
+
+  // Create spreadsheet if requested
```
Comment (chrismohrii) [inline capstone/src/main/webapp/sheetsScript.js:86]: Could the rest of the code in this method be made its own method? This one is getting a little long and I'm not sure if it's still related to the name `gatherSheetsInput`

File: capstone/src/main/webapp/sheetsScript.js [hunk 2]
```diff
@@ -62,29 +64,272 @@ async function gatherSheetsInput() {
...
+
+    const newSheetId = await createSheet(title);
+
+    // Check if user has entered a custom range
+    const userRangeElement = document.getElementById('sheets-range-yes-no');
+    if (userRangeElement != null && userRangeElement.value === 'yes') {
+      // Check if this is a valid Sheets range
+      const rangeElement = document.getElementById('sheets-range');
+      if (rangeElement != null) {
+        const rangeRegEx = rangeElement.value.match(/(.+)(![\w]*[\d]*(:[\w]*[\d]*)?)/);
```
Comment (chrismohrii) [inline capstone/src/main/webapp/sheetsScript.js]: nit: Could we put a comment for this regexp?

File: capstone/src/main/webapp/sheetsScript.js [hunk 3]
```diff
@@ -62,29 +64,272 @@ async function gatherSheetsInput() {
...
+ * and returns 0 if not found
+ */
+async function getSheetId(id, name) {
+  const response = await gapi.client.sheets.spreadsheets.get({
+    spreadsheetId: id,
+  });
+  const sheets = await response.result.sheets;
+  for (const sheet of sheets) {
+    if (sheet.properties.title === name)
+    return sheet.properties.sheetId;
```
Comment (chrismohrii) [inline capstone/src/main/webapp/sheetsScript.js]: nit: Should there be an extra indent here?

File: capstone/src/main/webapp/sheetsScript.js [hunk 4]
```diff
@@ -62,29 +64,272 @@ async function gatherSheetsInput() {
...
+  });
+  return await response.result.spreadsheetId;
+}
+
+/**
+ * Make the new spreadsheet contain the same sheet names
+ * as the one being analyzed
+ */
+async function preprocessSheet(id, sheetNames) {
+  var requests = [];
```
Comment (chrismohrii) [inline capstone/src/main/webapp/sheetsScript.js]: nit: `var` -> `let` and on line 276

File: capstone/src/main/webapp/sheetsScript.js [hunk 5]
```diff
@@ -62,29 +64,272 @@ async function gatherSheetsInput() {
...
+            midpoint: { color: { red: 1, green: 0.839, blue: 0.4, alpha: 1 }, type: 'PERCENTILE', value: '50' },
+            maxpoint: { color: { red: 0.902, green: 0.486, blue: 0.451, alpha: 1}, type: 'NUMBER', value: '1' }
+          }
+        },
+        index: 0
+      }}];
+
+  await gapi.client.sheets.spreadsheets.batchUpdate({
+    spreadsheetId: spreadsheetId,
+    resource: { requests }
```
Comment (chrismohrii) [inline capstone/src/main/webapp/sheetsScript.js:319]: nit: Is it possible to remove the brackets?

**Review by chrismohrii — APPROVED** [review]

File: capstone/src/main/webapp/sheetsScript.js (hunk 1 above)
Comment (danieldortega77) [reply capstone/src/main/webapp/sheetsScript.js:86]: Yes, sounds good. This is partly because I wanted the functionality to be working first before trying to separate out the code, but I can do that now.

File: capstone/src/main/webapp/sheetsScript.js (hunk 2 above)
Comment (danieldortega77) [reply capstone/src/main/webapp/sheetsScript.js]: Will do.

File: capstone/src/main/webapp/sheetsScript.js (hunk 3 above)
Comment (danieldortega77) [reply capstone/src/main/webapp/sheetsScript.js]: Yes - good catch

File: capstone/src/main/webapp/sheetsScript.js (hunk 4 above)
Comment (danieldortega77) [reply capstone/src/main/webapp/sheetsScript.js]: Done

File: capstone/src/main/webapp/sheetsScript.js (hunk 5 above)
Comment (danieldortega77) [reply capstone/src/main/webapp/sheetsScript.js:319]: Not that I know of - the API takes an object with an array of requests

File: capstone/src/main/webapp/sheetsScript.js [hunk 6]
```diff
@@ -62,29 +64,282 @@ async function gatherSheetsInput() {
...
+
+/**
+ * Returns all text in a Google Sheet
+ */
+function getTextFromSheet(sheet) {
+  if (sheet.values.length > 0) {
     let text = '';
-    for (i = 0; i < range.values.length; i++) {
-      let row = range.values[i];
+    for (i = 0; i < sheet.values.length; i++) {
```
Comment (olu144) [inline capstone/src/main/webapp/sheetsScript.js]: nit: lets make this for (let i = 0;.......)

**Review by olu144 — APPROVED** [review]
Looks Good!
I notice your color coding is done on a gradient (there are several shades of green, yellow and red used). We recently discussed switching to a "purple based" color coding system. Do you feel the Sheets output should follow suite or do you think it should be an exception as it is "special" in a sense?

File: capstone/src/main/webapp/sheetsScript.js (hunk 6 above)
Comment (danieldortega77) [reply capstone/src/main/webapp/sheetsScript.js]: Done!

**Comment by danieldortega77** [comment]
It wouldn't be hard to change the color of the gradient if we want to. I don't have a strong preference either way, but since the Sheet is separate from the website I think we can get away with a difference in style, at least slightly.

File: capstone/src/main/webapp/sheetsScript.js [hunk 7]
```diff
@@ -62,29 +64,282 @@ async function gatherSheetsInput() {
...
+    }
+  }
+
+  for (const name of sheetNames) {
+    const sheet = await getSpreadsheet(id, name, range);
+    const text = await getTextFromSheet(sheet);
+    totalText += text + '\n';
+  }
+
+  await handleInput(totalText, langElement.value, requestedAttributes, tokenizer);
```
Comment (dslucas) [inline capstone/src/main/webapp/sheetsScript.js:78]: Do we still want this call to `handleInput()`? I don't see it defined.

File: capstone/src/main/webapp/sheetsScript.js [hunk 8]
```diff
@@ -16,11 +16,11 @@
 const API_KEY = 'SHEETS_API_KEY'; // TODO: Create Java servlet to return key
 
 // Array of API discovery doc URLs for APIs
-const DISCOVERY_DOCS = ["https://sheets.googleapis.com/$discovery/rest?version=v4"];
+const DISCOVERY_DOCS = ['https://sheets.googleapis.com/$discovery/rest?version=v4'];
 
 // Authorization scopes required by the API; multiple scopes can be
 // included, separated by spaces.
-const SCOPES = "https://www.googleapis.com/auth/spreadsheets";
+const SCOPES = 'https://www.googleapis.com/auth/spreadsheets';
```
Comment (dslucas) [inline capstone/src/main/webapp/sheetsScript.js]: nit: Consider renaming this to something like `SHEETS_SCOPE` or `API_SCOPE` if we want to keep this as just a single string, rather than a list of scopes.

File: capstone/src/main/webapp/sheetsScript.js [hunk 9]
```diff
@@ -62,29 +64,282 @@ async function gatherSheetsInput() {
...
+    let title = 'Perspective Output';
+    const titleElement = document.getElementById('sheets-title');
+    if (titleElement != null && titleElement.value != '') {
+      title = titleElement.value;
+    }
+
+    const newSheetId = await createSheet(title);
+
+    // Make the new spreadsheet mirror sheet names of the input
+    if (sheetNames[0] != 'Sheet1') {
```
Comment (dslucas) [inline capstone/src/main/webapp/sheetsScript.js]: Should we be checking all the sheet names? In addition to just the first?

File: capstone/src/main/webapp/sheetsScript.js [hunk 10]
```diff
@@ -62,29 +64,282 @@ async function gatherSheetsInput() {
...
+          promises.push(callPerspective(row[j], lang, requestedAttributes));
+          body.push([row[j]]);
+        }
+      }
+    }
+
+    // Evaluate Perspective calls
+    const resolvedResponses = await Promise.all(promises);
+    for (index = 0; index < promises.length; index++) {
+      if (resolvedResponses[index].attributeScores) {
```
Comment (dslucas) [inline capstone/src/main/webapp/sheetsScript.js:265]: Does this check mean that comments that got no scores (e.g. from a language detection failure) will not show up on the output? If so, we may want to consider making that explicit or including some sort of "Unable to score" message in the output sheet.

File: capstone/src/main/webapp/sheetsScript.js [hunk 11]
```diff
@@ -62,29 +64,282 @@ async function gatherSheetsInput() {
...
 
-  const text = await getTextFromSheet(id);
-  handleInput(text, langElement.value, requestedAttributes, tokenizer);
+  // Show general output
+  let sheetNames = await getSheetNames(id);
+  let totalText = '';
+  let range = '!A1:YY';
+
+  // Check if user has entered a custom range
+  const userRangeElement = document.getElementById('sheets-range-yes-no');
```
Comment (dborkan) [inline capstone/src/main/webapp/sheetsScript.js]: This code is a little hard to follow because above you do `let sheetNames = await getSheetNames(id);`, but then you don't use that value right away and instead possibly overwrite sheetNames based on the user input.

Would it maybe make sense to combine all of this logic for the sheetNames and range into some separate function?  Also maybe if the user specifies something in the `sheets-range` element, then you don't need to make the sheets API request, which might be a little more efficient?

File: capstone/src/main/webapp/sheetsScript.js (hunk 7 above)
Comment (danieldortega77) [reply capstone/src/main/webapp/sheetsScript.js:78]: This references the `handleInput()` from `script.js`, which creates the bar graph, detailed output, and Datamuse area. We can still change the name if necessary, but I didn't want to touch the existing `script.js`.

File: capstone/src/main/webapp/sheetsScript.js (hunk 8 above)
Comment (danieldortega77) [reply capstone/src/main/webapp/sheetsScript.js]: Sounds good - I'll do that.

File: capstone/src/main/webapp/sheetsScript.js (hunk 9 above)
Comment (danieldortega77) [reply capstone/src/main/webapp/sheetsScript.js]: That's correct; I made this as a quick fix to an error but forgot to account for all cases. I've just changed it.

File: capstone/src/main/webapp/sheetsScript.js (hunk 10 above)
Comment (danieldortega77) [reply capstone/src/main/webapp/sheetsScript.js:265]: Yes, that's what this checks. I'll add a message to the output if an error occurs.

File: capstone/src/main/webapp/sheetsScript.js (hunk 11 above)
Comment (danieldortega77) [reply capstone/src/main/webapp/sheetsScript.js]: That makes sense. I'll refactor this into a separate function that returns `sheetNames` and `range` and reorganize the logic.

**Review by dborkan — APPROVED** [review]
LGTM

File: capstone/src/main/webapp/sheetsScript.js [hunk 12]
```diff
@@ -62,29 +62,302 @@ async function gatherSheetsInput() {
     }
   }
 
-  const text = await getTextFromSheet(id);
-  handleInput(text, langElement.value, requestedAttributes, tokenizer);
+  // Get the name and range of sheets to be analyzed
+  const response = await getSheetNamesAndRange(id);
+  const sheetNames = await response.sheetNames;
+  const range = await response.range;
```
Comment (dslucas) [inline capstone/src/main/webapp/sheetsScript.js]: nit: I think you may be able to omit these `await`s because you already `await` for the response in the line above.

File: capstone/src/main/webapp/sheetsScript.js [hunk 13]
```diff
@@ -62,29 +62,302 @@ async function gatherSheetsInput() {
...
+
+  // Show general output
+  let totalText = '';
+  for (const name of sheetNames) {
+    const sheet = await getSpreadsheet(id, name, range);
+    const text = await getTextFromSheet(sheet);
```
Comment (dslucas) [inline capstone/src/main/webapp/sheetsScript.js]: nit: Same comment about `await` here.

File: capstone/src/main/webapp/sheetsScript.js [hunk 14]
```diff
@@ -62,29 +62,302 @@ async function gatherSheetsInput() {
...
+      if (rangeRegEx != null) {
+        return { 'sheetNames': [rangeRegEx[1]], 'range': rangeRegEx[2] };
+      } else {
+        alert('Please input a range in a valid format.');
+      }
+    }
+  }
+
+  // Otherwise, get all sheets
+  let sheetNames = await getSheetNames(id);
```
Comment (dslucas) [inline capstone/src/main/webapp/sheetsScript.js]:
```suggestion
  const sheetNames = await getSheetNames(id);
```

**Review by dslucas — APPROVED** [review]

File: capstone/src/main/webapp/sheetsScript.js (hunk 12 above)
Comment (danieldortega77) [reply capstone/src/main/webapp/sheetsScript.js]: Sounds good - I'll remove them.

File: capstone/src/main/webapp/sheetsScript.js (hunk 13 above)
Comment (danieldortega77) [reply capstone/src/main/webapp/sheetsScript.js]: Done

File: capstone/src/main/webapp/sheetsScript.js (hunk 14 above)
Comment (danieldortega77) [reply capstone/src/main/webapp/sheetsScript.js]: Good catch - I changed it now.
